wtransport/src/endpoint.rs

```diff
@@ -564,18 +565,66 @@ where
 type DynFutureIncomingSession =
     dyn Future<Output = Result<SessionRequest, ConnectionError>> + Send + Sync;
 
-/// [`Future`] for an in-progress incoming connection attempt.
+/// [`IntoFuture`] for an in-progress incoming connection attempt.
 ///
 /// Created by [`Endpoint::accept`].
-pub struct IncomingSession(Pin<Box<DynFutureIncomingSession>>);
+pub struct IncomingSession(quinn::Incoming);
 
 impl IncomingSession {
-    fn new(quic_connecting: quinn::Connecting) -> Self {
-        Self(Box::pin(Self::accept(quic_connecting)))
+    /// The peer's UDP address.
+    pub fn remote_address(&self) -> SocketAddr {
+        self.0.remote_address()
+    }
+
+    /// Whether the socket address that is initiating this connection has been validated.
+    ///
+    /// This means that the sender of the initial packet has proved that they can receive traffic
+    /// sent to `self.remote_address()`.
+    pub fn remote_address_validated(&self) -> bool {
+        self.0.remote_address_validated()
+    }
+
+    /// Respond with a retry packet, requiring the client to retry with address validation
+    ///
+    /// # Panics
+    ///
+    /// If `remote_address_validated()` is true.
+    pub fn retry(self) {
+        self.0.retry().expect("remote address already verified");
+    }
+
+    /// Reject this incoming connection attempt.
+    pub fn refuse(self) {
+        self.0.refuse();
+    }
+
+    /// Ignore this incoming connection attempt, not sending any packet in response.
+    pub fn ignore(self) {
+        self.0.ignore();
+    }
+}
+
+impl IntoFuture for IncomingSession {
+    type IntoFuture = IncomingSessionFuture;
+    type Output = Result<SessionRequest, ConnectionError>;
+
+    fn into_future(self) -> Self::IntoFuture {
+        IncomingSessionFuture::new(self.0)
+    }
+}
+
+/// [`Future`] for an in-progress incoming connection attempt.
+///
+/// Created by awaiting an [`IncomingSession`]
+pub struct IncomingSessionFuture(Pin<Box<DynFutureIncomingSession>>);
```

Comment:
Previously, one could use `quinn`'s `max_concurrent_connections` and `use_retry` to mitigate DoS, but those were removed.

I experimented with re-implementing them in `wtransport` here: https://github.com/finnbear/wtransport/commit/5dad5aac94a151047602e4a4e36603ff2ea2390d

But `IncomingSession` methods are much more flexible:
- can decide based on IP address
- can randomly `retry` based on load i.e. https://www.ietf.org/archive/id/draft-ietf-quic-retry-offload-00.html